Sources/ContextPropagation/Instrumentation/InstrumentationMiddleware.swift

```diff
@@ -18,6 +18,10 @@ public struct InstrumentationMiddleware<InjectInto, ExtractFrom>: Instrumentatio
         self.inject = inject
     }
 
+    public init<M: InstrumentationMiddlewareProtocol>(_ middleware: M) where M.InjectInto == InjectInto, M.ExtractFrom == ExtractFrom {
+        self = .init(extract: middleware.extract, inject: middleware.inject)
```

Comment:
This init is a bit meh, we'd force everyone who has to perform the wrapping to do the "give me the callbacks" manually which gets annoying, consider the following init implementation:

```swift
public struct InstrumentationMiddleware<InjectInto, ExtractFrom>: InstrumentationMiddlewareProtocol {
    private let _extract: (ExtractFrom, inout Context) -> Void
    private let _inject: (Context, inout InjectInto) -> Void

    public init<Middleware>(_ middleware: Middleware)
        where
        Middleware: InstrumentationMiddlewareProtocol,
        Middleware.ExtractFrom == Self.ExtractFrom,
        Middleware.InjectInto == Self.InjectInto {
            _extract = { from, context in middleware.extract(from: from, into: &context) }
            _inject = { context, into in middleware.inject(from: context, into: &into) }
    }

    public func extract(from: ExtractFrom, into context: inout Context) {
        _extract(from, &context)
    }

    public func inject(from context: Context, into: inout InjectInto) {
        _inject(context, &into)
    }
}

```

Then the callsite here can be

```suggestion
        self = .init(middleware)
```

---

we could also avoid "double wrapping" with

```
        if let alreadyWrapped = middleware as? InstrumentationMiddleware<InjectInto, ExtractFrom> {
            self = alreadyWrapped
        } else {
...
```

or making the wrapping a function on the protocol `as...: InstrumentationMiddleware { return self }` (if already this), and `: InstrumentationMiddleware { .init(self) }` by default on the protocol... 

But that's not really necessary right now; though let's change the init so it takes a middleware and not the functions though 👍